torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+                               : ir_builder.namedSetExpr(name, val);
+  auto alloc = ir_builder.create<kir::Allocate>(
+      namedScalar, MemoryType::Local, ir_builder.one());
+  body.push_back(alloc);
+  body.push_back(namedScalar->definition());
+  return namedScalar;
+}
+
+kir::ForLoop* handleMisalignedVectorization(
+    std::vector<kir::ForLoop*> loop_structure,
+    const kir::ForLoop* for_loop) {
+  // for_loop body contains allocate, read, compute, write operations
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+
+  // create new base For-Loop
+  const auto new_loop = ir_builder.create<kir::ForLoop>(
+      for_loop->index(), for_loop->iter_domain(), for_loop->extent());
+
+  // Find child For-Loops and add remaining expressions to base For-Loop
+  auto child_loops = parseVectorizedForLoop(for_loop, new_loop);
+
+  // Find the first vectorize set - either read or write
+  auto vec_expr = findVectorizedSet(loop_structure, child_loops);
+  TORCH_INTERNAL_ASSERT(vec_expr != nullptr);
+  TORCH_INTERNAL_ASSERT(vec_expr->outputs().front()->isA<kir::TensorView>());
+  TORCH_INTERNAL_ASSERT(vec_expr->inputs().front()->isA<kir::TensorView>());
+
+  auto out_tv = vec_expr->outputs().front()->as<kir::TensorView>();
+  auto in_tv = vec_expr->inputs().front()->as<kir::TensorView>();
+
+  // It is assumed that either of input and output is on global memory
+  TORCH_INTERNAL_ASSERT(
+      out_tv->memoryType() == MemoryType::Global ||
+          in_tv->memoryType() == MemoryType::Global,
+      "Either input or output tensor must be on global memory.");
+  // However, not both of them
+  TORCH_INTERNAL_ASSERT(
+      !(out_tv->memoryType() == MemoryType::Global &&
+        in_tv->memoryType() == MemoryType::Global),
+      "Both input and output tensors are on global memory.");
+  // Must be either global or local
+  TORCH_INTERNAL_ASSERT(
+      (out_tv->memoryType() == MemoryType::Global ||
+       out_tv->memoryType() == MemoryType::Local),
+      "Invalid memory type of output tensor");
+  TORCH_INTERNAL_ASSERT(
+      (in_tv->memoryType() == MemoryType::Global ||
+       in_tv->memoryType() == MemoryType::Local),
+      "Invalid memory type of input tensor");
+
+  // TensorView on global memory. This is the tensor that may have
+  // a non-aligned base address.
+  auto global_tv =
+      (out_tv->memoryType() == MemoryType::Global) ? out_tv : in_tv;
+
+  // TensorView with the misaligned vec iterDomain. It is the consumer
+  // of vectorized load or the producer of vectorized store. It is
+  // assumed that when the output TV is not on global memory, this
+  // expression is a vectorized load, so the output TV is vec_tv.
+  // TODO: Check vec_tv has indeed MisalignedVectorize parallel type.
+  auto vec_tv = (out_tv->memoryType() != MemoryType::Global) ? out_tv : in_tv;
+
+  auto pred = PredicateCompute::getInlinePredicate(
+      vec_expr, loop_structure, nullptr, false, true);
+  if (pred == nullptr) {
```

Comment:
Why is getInlinePredicate returning nullptr?

Reply:
I'm passing `nullptr `for the `thread_pred `argument. getInlinePredicate can return 'thread_pred' by itself.
If `thread_pred `isn't `nullptr`, then it will be added to the predicate.

Reply:
Provide `thread_pred` as `true`. `getInlinePredicate` should not return a `nullptr`

Reply:
Then, why do we have a check like this?
https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/predicate_compute.cpp#L251-L254